File: examples/typescript/07-memory/tests/memory-provider.test.ts
/**
 * Unit tests for the `MarkdownMemoryProvider` shipped by example 07.
 *
 * They are driven directly over a temp `memory.md` file (no model, no Ollama),
 * proving each behavior the example demonstrates:
 *   - a missing file reads as empty.
 *   - append → get round-trips an entry, and writes real readable markdown.
 *   - multi-line content round-trips.
 *   - scope filtering isolates scopes.
 *   - session filtering isolates sessions.
 *   - get returns newest-first by timestamp.
 *   - limit takes the most-recent N.
 *   - a hand-edited file is tolerated (prose + extra headings ignored).
 *   - composition: the markdown provider fills the memory slot of a full
 *     StorageProvider, NoOp fills the rest.
 */

import {
  existsSync,
  mkdtempSync,
  readFileSync,
  rmSync,
  writeFileSync,
} from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";

import {
  SessionId,
  memory as coreMemory,
  storage as coreStorage,
} from "@spore/core";
import { afterEach, beforeEach, describe, expect, it } from "vitest";

import { MarkdownMemoryProvider } from "../src/memory-provider.js";

type MemoryEntry = coreStorage.MemoryEntry;

const { Timestamp } = coreMemory;

function entry(role: string, content: string, ts: string): MemoryEntry {
  return { role, content, timestamp: Timestamp.of(ts), metadata: {} };
}

function sid(): SessionId {
  return SessionId.of("s1");
}

let dir: string;
let path: string;

beforeEach(() => {
  dir = mkdtempSync(join(tmpdir(), "spore07-"));
  path = join(dir, "memory.md");
});

afterEach(() => {
  rmSync(dir, { recursive: true, force: true });
});

describe("MarkdownMemoryProvider", () => {
  it("reads a missing file as empty", async () => {
    const provider = new MarkdownMemoryProvider(path);
    const got = await provider.getMemories("project", sid(), 50);
    expect(got).toEqual([]);
  });

  it("round-trips an appended entry and writes readable markdown", async () => {
    const provider = new MarkdownMemoryProvider(path);
    await provider.appendMemory(
      "project",
      sid(),
      entry(
        "assistant",
        "Postgres is the system of record.",
        "2026-06-02T10:00:00Z",
      ),
    );

    const got = await provider.getMemories("project", sid(), 50);
    expect(got).toHaveLength(1);
    expect(got[0]!.role).toBe("assistant");
    expect(got[0]!.content).toBe("Postgres is the system of record.");
    expect(got[0]!.timestamp.asString()).toBe("2026-06-02T10:00:00Z");

    // The artifact is real, readable markdown on disk.
    const raw = readFileSync(path, "utf8");
    expect(raw).toContain("## [project] [s1] 2026-06-02T10:00:00Z — assistant");
    expect(raw).toContain("Postgres is the system of record.");
  });

  it("round-trips multi-line content", async () => {
    const provider = new MarkdownMemoryProvider(path);
    await provider.appendMemory(
      "project",
      sid(),
      entry(
        "assistant",
        "line one\nline two\n\nline four",
        "2026-06-02T10:00:00Z",
      ),
    );
    const got = await provider.getMemories("project", sid(), 50);
    expect(got[0]!.content).toBe("line one\nline two\n\nline four");
  });

  it("isolates scopes", async () => {
    const provider = new MarkdownMemoryProvider(path);
    await provider.appendMemory(
      "project",
      sid(),
      entry("user", "proj", "2026-06-02T10:00:00Z"),
    );
    await provider.appendMemory(
      "user",
      sid(),
      entry("user", "usr", "2026-06-02T10:00:01Z"),
    );

    const proj = await provider.getMemories("project", sid(), 50);
    expect(proj).toHaveLength(1);
    expect(proj[0]!.content).toBe("proj");

    const usr = await provider.getMemories("user", sid(), 50);
    expect(usr).toHaveLength(1);
    expect(usr[0]!.content).toBe("usr");
  });

  it("isolates sessions", async () => {
    const provider = new MarkdownMemoryProvider(path);
    const a = SessionId.of("alpha");
    const b = SessionId.of("beta");
    await provider.appendMemory(
      "project",
      a,
      entry("user", "from-alpha", "2026-06-02T10:00:00Z"),
    );
    await provider.appendMemory(
      "project",
      b,
      entry("user", "from-beta", "2026-06-02T10:00:01Z"),
    );

    const gotA = await provider.getMemories("project", a, 50);
    expect(gotA).toHaveLength(1);
    expect(gotA[0]!.content).toBe("from-alpha");
  });

  it("returns newest-first by timestamp", async () => {
    const provider = new MarkdownMemoryProvider(path);
    // Appended out of timestamp order on purpose.
    await provider.appendMemory(
      "project",
      sid(),
      entry("user", "middle", "2026-06-02T11:00:00Z"),
    );
    await provider.appendMemory(
      "project",
      sid(),
      entry("user", "oldest", "2026-06-02T10:00:00Z"),
    );
    await provider.appendMemory(
      "project",
      sid(),
      entry("user", "newest", "2026-06-02T12:00:00Z"),
    );

    const got = await provider.getMemories("project", sid(), 50);
    expect(got.map((e) => e.content)).toEqual(["newest", "middle", "oldest"]);
  });

  it("limit takes the most-recent entries", async () => {
    const provider = new MarkdownMemoryProvider(path);
    for (let i = 0; i < 5; i++) {
      await provider.appendMemory(
        "project",
        sid(),
        entry("user", `e${i}`, `2026-06-02T10:00:0${i}Z`),
      );
    }
    const got = await provider.getMemories("project", sid(), 2);
    expect(got.map((e) => e.content)).toEqual(["e4", "e3"]);
  });

  it("tolerates a hand-edited file", async () => {
    // A file a human authored/edited: prose before the first header, an extra
    // heading, blank lines, and a normal entry block.
    const hand =
      "# My Notes\n\n" +
      "Some rambling prose that is not an entry.\n\n" +
      "## [project] [s1] 2026-06-02T09:00:00Z — user\n\n" +
      "Hand-written fact about Ironwood.\n\n" +
      "## A non-entry heading the human added\n\n" +
      "more prose\n";
    writeFileSync(path, hand);

    const provider = new MarkdownMemoryProvider(path);
    let got = await provider.getMemories("project", sid(), 50);
    expect(got).toHaveLength(1);
    expect(got[0]!.content).toBe("Hand-written fact about Ironwood.");

    // And we can still append on top of the hand-edited file.
    await provider.appendMemory(
      "project",
      sid(),
      entry("assistant", "appended", "2026-06-02T10:00:00Z"),
    );
    got = await provider.getMemories("project", sid(), 50);
    expect(got).toHaveLength(2);
    expect(got[0]!.content).toBe("appended"); // newest-first
  });

  it("composes into the memory slot of a full StorageProvider", async () => {
    const provider = new MarkdownMemoryProvider(path);
    const storage = provider.intoStorageProvider();

    // The memory slot is the markdown provider; round-trips through it.
    await storage
      .memory()
      .appendMemory(
        "project",
        sid(),
        entry("user", "via-seam", "2026-06-02T10:00:00Z"),
      );
    const got = await storage.memory().getMemories("project", sid(), 50);
    expect(got).toHaveLength(1);
    expect(got[0]!.content).toBe("via-seam");

    // The other three domains are no-ops: a run read returns nothing.
    expect(await storage.run().get(sid(), "k")).toBeUndefined();

    // The artifact exists on disk.
    expect(existsSync(path)).toBe(true);
  });
});

File: examples/go/07-memory/memory_provider_test.go
package main

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"testing"

	sporecore "github.com/squirrelsoft-dev/spore-core/go/spore-core"
	"github.com/squirrelsoft-dev/spore-core/go/spore-core/storage"
)

func tempPath(t *testing.T) string {
	t.Helper()
	return filepath.Join(t.TempDir(), "memory.md")
}

func sid() storage.SessionID { return storage.SessionID("s1") }

func entry(role, content, ts string) storage.MemoryEntry {
	return storage.NewMemoryEntry(role, content, storage.Timestamp(ts))
}

func contents(entries []storage.MemoryEntry) []string {
	out := make([]string, len(entries))
	for i, e := range entries {
		out[i] = e.Content
	}
	return out
}

func mustAppend(t *testing.T, p *MarkdownMemoryProvider, scope storage.StorageScope, s storage.SessionID, e storage.MemoryEntry) {
	t.Helper()
	if err := p.AppendMemory(context.Background(), scope, s, e); err != nil {
		t.Fatalf("AppendMemory: %v", err)
	}
}

func mustGet(t *testing.T, p *MarkdownMemoryProvider, scope storage.StorageScope, s storage.SessionID, limit int) []storage.MemoryEntry {
	t.Helper()
	got, err := p.GetMemories(context.Background(), scope, s, limit)
	if err != nil {
		t.Fatalf("GetMemories: %v", err)
	}
	return got
}

func TestMissingFileReadsEmpty(t *testing.T) {
	p := NewMarkdownMemoryProvider(tempPath(t))
	got := mustGet(t, p, storage.StorageScopeProject, sid(), 50)
	if len(got) != 0 {
		t.Fatalf("want empty, got %d entries", len(got))
	}
}

func TestAppendThenGetRoundtrips(t *testing.T) {
	path := tempPath(t)
	p := NewMarkdownMemoryProvider(path)
	mustAppend(t, p, storage.StorageScopeProject, sid(),
		entry("assistant", "Postgres is the system of record.", "2026-06-02T10:00:00Z"))

	got := mustGet(t, p, storage.StorageScopeProject, sid(), 50)
	if len(got) != 1 {
		t.Fatalf("want 1 entry, got %d", len(got))
	}
	if got[0].Role != "assistant" {
		t.Fatalf("role: %q", got[0].Role)
	}
	if got[0].Content != "Postgres is the system of record." {
		t.Fatalf("content: %q", got[0].Content)
	}
	if string(got[0].Timestamp) != "2026-06-02T10:00:00Z" {
		t.Fatalf("timestamp: %q", got[0].Timestamp)
	}

	// The artifact is real, readable markdown on disk.
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read file: %v", err)
	}
	if !strings.Contains(string(raw), "## [project] [s1] 2026-06-02T10:00:00Z — assistant") {
		t.Fatalf("missing header in file:\n%s", raw)
	}
	if !strings.Contains(string(raw), "Postgres is the system of record.") {
		t.Fatalf("missing content in file:\n%s", raw)
	}
}

func TestMultilineContentRoundtrips(t *testing.T) {
	p := NewMarkdownMemoryProvider(tempPath(t))
	mustAppend(t, p, storage.StorageScopeProject, sid(),
		entry("assistant", "line one\nline two\n\nline four", "2026-06-02T10:00:00Z"))

	got := mustGet(t, p, storage.StorageScopeProject, sid(), 50)
	if len(got) != 1 {
		t.Fatalf("want 1 entry, got %d", len(got))
	}
	if got[0].Content != "line one\nline two\n\nline four" {
		t.Fatalf("content: %q", got[0].Content)
	}
}

func TestScopeFilteringIsolatesScopes(t *testing.T) {
	p := NewMarkdownMemoryProvider(tempPath(t))
	mustAppend(t, p, storage.StorageScopeProject, sid(),
		entry("user", "proj", "2026-06-02T10:00:00Z"))
	mustAppend(t, p, storage.StorageScopeUser, sid(),
		entry("user", "usr", "2026-06-02T10:00:01Z"))

	proj := mustGet(t, p, storage.StorageScopeProject, sid(), 50)
	if len(proj) != 1 || proj[0].Content != "proj" {
		t.Fatalf("project scope: %v", contents(proj))
	}
	usr := mustGet(t, p, storage.StorageScopeUser, sid(), 50)
	if len(usr) != 1 || usr[0].Content != "usr" {
		t.Fatalf("user scope: %v", contents(usr))
	}
}

func TestSessionFilteringIsolatesSessions(t *testing.T) {
	p := NewMarkdownMemoryProvider(tempPath(t))
	a := storage.SessionID("alpha")
	b := storage.SessionID("beta")
	mustAppend(t, p, storage.StorageScopeProject, a,
		entry("user", "from-alpha", "2026-06-02T10:00:00Z"))
	mustAppend(t, p, storage.StorageScopeProject, b,
		entry("user", "from-beta", "2026-06-02T10:00:01Z"))

	gotA := mustGet(t, p, storage.StorageScopeProject, a, 50)
	if len(gotA) != 1 || gotA[0].Content != "from-alpha" {
		t.Fatalf("alpha session: %v", contents(gotA))
	}
}

func TestGetReturnsNewestFirst(t *testing.T) {
	p := NewMarkdownMemoryProvider(tempPath(t))
	// Appended out of timestamp order on purpose.
	mustAppend(t, p, storage.StorageScopeProject, sid(),
		entry("user", "middle", "2026-06-02T11:00:00Z"))
	mustAppend(t, p, storage.StorageScopeProject, sid(),
		entry("user", "oldest", "2026-06-02T10:00:00Z"))
	mustAppend(t, p, storage.StorageScopeProject, sid(),
		entry("user", "newest", "2026-06-02T12:00:00Z"))

	got := mustGet(t, p, storage.StorageScopeProject, sid(), 50)
	want := []string{"newest", "middle", "oldest"}
	if fmt.Sprint(contents(got)) != fmt.Sprint(want) {
		t.Fatalf("order: got %v, want %v", contents(got), want)
	}
}

func TestLimitTakesMostRecent(t *testing.T) {
	p := NewMarkdownMemoryProvider(tempPath(t))
	for i := 0; i < 5; i++ {
		mustAppend(t, p, storage.StorageScopeProject, sid(),
			entry("user", fmt.Sprintf("e%d", i), fmt.Sprintf("2026-06-02T10:00:0%dZ", i)))
	}
	got := mustGet(t, p, storage.StorageScopeProject, sid(), 2)
	want := []string{"e4", "e3"}
	if fmt.Sprint(contents(got)) != fmt.Sprint(want) {
		t.Fatalf("limit: got %v, want %v", contents(got), want)
	}
}

func TestToleratesHandEditedFile(t *testing.T) {
	path := tempPath(t)
	// A file a human authored/edited: prose before the first header, an extra
	// heading, blank lines, and a normal entry block.
	hand := "# My Notes\n\nSome rambling prose that is not an entry.\n\n" +
		"## [project] [s1] 2026-06-02T09:00:00Z — user\n\n" +
		"Hand-written fact about Ironwood.\n\n" +
		"## A non-entry heading the human added\n\n" +
		"more prose\n"
	if err := os.WriteFile(path, []byte(hand), 0o644); err != nil {
		t.Fatalf("write file: %v", err)
	}
	p := NewMarkdownMemoryProvider(path)

	got := mustGet(t, p, storage.StorageScopeProject, sid(), 50)
	if len(got) != 1 || got[0].Content != "Hand-written fact about Ironwood." {
		t.Fatalf("hand-edited read: %v", contents(got))
	}

	// And we can still append on top of the hand-edited file.
	mustAppend(t, p, storage.StorageScopeProject, sid(),
		entry("assistant", "appended", "2026-06-02T10:00:00Z"))
	got = mustGet(t, p, storage.StorageScopeProject, sid(), 50)
	if len(got) != 2 {
		t.Fatalf("want 2 entries after append, got %d", len(got))
	}
	if got[0].Content != "appended" { // newest-first
		t.Fatalf("newest after append: %q", got[0].Content)
	}
}

func TestComposesIntoStorageProviderMemorySlot(t *testing.T) {
	p := NewMarkdownMemoryProvider(tempPath(t))
	sp := p.IntoStorageProvider()

	// The memory slot is the markdown provider; round-trips through it.
	if err := sp.Memory().AppendMemory(context.Background(), storage.StorageScopeProject, sid(),
		entry("user", "via-seam", "2026-06-02T10:00:00Z")); err != nil {
		t.Fatalf("AppendMemory via seam: %v", err)
	}
	got, err := sp.Memory().GetMemories(context.Background(), storage.StorageScopeProject, sid(), 50)
	if err != nil {
		t.Fatalf("GetMemories via seam: %v", err)
	}
	if len(got) != 1 || got[0].Content != "via-seam" {
		t.Fatalf("via-seam read: %v", contents(got))
	}

	// The other three domains are no-ops: a run read returns nothing.
	v, found, err := sp.Run().Get(context.Background(), sid(), "k")
	if err != nil {
		t.Fatalf("Run().Get: %v", err)
	}
	if found || v != nil {
		t.Fatalf("expected no-op run store, got found=%v v=%s", found, v)
	}
}

func TestMergedReadUnionsUserAndProject(t *testing.T) {
	p := NewMarkdownMemoryProvider(tempPath(t))
	mustAppend(t, p, storage.StorageScopeProject, sid(),
		entry("assistant", "proj", "2026-06-02T10:00:00Z"))
	mustAppend(t, p, storage.StorageScopeUser, sid(),
		entry("assistant", "usr", "2026-06-02T11:00:00Z"))
	mustAppend(t, p, storage.StorageScopeLocal, sid(),
		entry("assistant", "loc", "2026-06-02T12:00:00Z"))

	// GetMemoriesMerged delegates to storage.MergeMemories: User ∪ Project,
	// newest-first, Local excluded.
	got, err := p.GetMemoriesMerged(context.Background(), sid(), 50)
	if err != nil {
		t.Fatalf("GetMemoriesMerged: %v", err)
	}
	want := []string{"usr", "proj"} // newest-first, no local
	if fmt.Sprint(contents(got)) != fmt.Sprint(want) {
		t.Fatalf("merged: got %v, want %v", contents(got), want)
	}
}

// Compile-time check that the example's session id helper returns the pinned id.
func TestPinnedSessionID(t *testing.T) {
	if session() != sporecore.SessionID("project-ironwood") {
		t.Fatalf("pinned session id changed: %q", session())
	}
}
